features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,329 @@
+operation would not trigger AFTER events.
 
-To acknowledge receipt of a webhook, webhook endpoint should return a 2xx HTTP status code. All response codes outside this range, including 3xx codes, will indicate to skygear the request is failed.
+### AFTER Events
 
-For some events, it may allow to webhook to modify changes (e.x. auth gear's before family hooks) when response code is 2xx. The behavior is defined by each gear, so will not be covered in this spec. Missing field in the return object means unmodified, so empty body is accepted. Another thing is gear will validate return object, if validation failed, gear should return an error to SDK.
+AFTER events would be delivered to web-hook handlers asynchronously after the
+operation is performed (i.e. commited into database).
 
-Take auth gear as an example, its before family hooks accept that a hook returns a user object to modify current user object.
+The time spent in an AFTER event delivery must not exceed 60 seconds, otherwise
+would be considered a failed delivery.
 
-```javascript=
-var express = require('express');
-var bodyParser = require('body-parser');
+All AFTER events with web-hook handler are presisted into database, with
+minimum retention period of 30 days.
 
-var app = express();
+The response body of web-hook handler is ignored.
 
-app.use(bodyParser.json());
+If any delivery failed, all deliveries will be retried after some time,
+regardless whether some deliveries may succeed. The retry will be performed
+with a variant of exponential back-off algorithm. If `Retry-After` HTTP header
+is present in the response, the delivery will not be retried before the
+specified time.
 
-app.post('/before_signup', function(req, res){
-    var user = req.body.data.user;
+If the delivery continue to fail after 3 days from the time of first attempted
+delivery, the event is marked as permanently failed and will not be retried
+automatically.
 
-    user.metadata.loveCat = true;
+## Mutations
 
-    res.status(200);
-    res.send(user);
-});
+BEFORE events allow web-hook handler to request mutation on some fields before
+committing to database.
+
+Web-hook handler cannot request mutation if the operation is disallowed.
+Web-hook handler should indicate the mutations to perform through its response,
+for example:
+```json
+{
+    "is_allowed": true,
+    "mutations": {
+        "metadata": {
+            "username": "test"
+        },
+        "is_verified": false
+    }
+}
 ```
 
-If webhook wants to include additional information for an error case, it could return with payload:
+- If a field is absent: no mutation would be performed for that field.
+- If a field is present: the field would be set to the provided value.
+
+If failed to perform the mutations, the operation would be failed.
+If the operation failed, the mutations are rollbacked and not visible.
+
+Mutations would not generate additional events (except `user_sync`).
```

Comment:
`user_sync` is an after event I think, so there are no mutation?

Reply:
Yes, I mean mutations would not generate additional events, however `user_sync` would still be generated.

Reply:
But logically `user_sync` have no mutation, so it is werid to say (except `user_sync`)?

Reply:
Yes I've further clarified the except clause in the spec.

Reply:
The clarification still read werid, readability and logically-wise seems doesn't make sense to specially mention `user_sync` which is an after async event with no mutation supports.

I guess you actually meant something like this?

> Mutations would not generate additional events unless users call other APIs to update the data independent from returning mutation data in the web hook.

Reply:
I'll remove that note instead, since `user_sync` is not generated based on mutations.